File: c/linked-list/linked_list.c
#include "linked_list.h"
#include <stdbool.h>
#include <stdlib.h>

// Represents a node of the doubly linked list.
// Stores a value and has a previous neighbor and a next neighbor.
typedef struct list_node {
  struct list_node *prev, *next;
  ll_data_t data;
} list_node_t;

// Represents a doubly linked list. Stores the count of elements, as well as a
// reference to the first and last nodes.
typedef struct list {
  struct list_node *first, *last;
  size_t count;
} list_t;

list_node_t *node_create(ll_data_t data, list_node_t *prev, list_node_t *next);
void node_destroy(list_node_t *node);
bool list_is_empty(const list_t *list);
bool list_is_single(const list_t *list);

// counts the items on a list
size_t list_count(const list_t *list) {
  if (list) {
    return list->count;
  }
  return 0;
}

// inserts item at back of a list
void list_push(list_t *list, ll_data_t item_data) {
  if (!list) {
    return;
  }
  list->count++;

  if (list_is_empty(list)) {
    list->last = node_create(item_data, NULL, NULL);
    list->first = list->last;
    return;
  }
  list->last->next = node_create(item_data, list->last, NULL);
  list->last = list->last->next;
}

// inserts item at front of a list
void list_unshift(list_t *list, ll_data_t item_data) {
  if (!list) {
    return;
  }
  list->count++;

  if (list_is_empty(list)) {
    list->first = node_create(item_data, NULL, NULL);
    list->last = list->first;
    return;
  }
  list->first->prev = node_create(item_data, NULL, list->first);
  list->first = list->first->prev;
}

// removes item from back of a list
ll_data_t list_pop(list_t *list) {
  if (!list || list_is_empty(list)) {
    return 0;
  }

  list->count--;

  list_node_t *popped = list->last;
  const ll_data_t popped_data = popped->data;
  // printf("Popping: %d\n", popped_data);

  if (list_is_single(list)) {
    list->first = NULL;
    list->last = NULL;
  } else {
    popped->prev->next = NULL;
    list->last = popped->prev;
  }

  node_destroy(popped);
  return popped_data;
}
// removes item from front of a list
ll_data_t list_shift(list_t *list) {
  if (!list || list_is_empty(list)) {
    return 0;
  }
  list->count--;

  list_node_t *shifted = list->first;
  const ll_data_t shifted_data = list->first->data;

  if (list_is_single(list)) {
    list->first = NULL;
    list->last = NULL;
  } else {
    shifted->next->prev = NULL;
    list->first = shifted->next;
  }

  node_destroy(shifted);
  return shifted_data;
}

// deletes a node that holds the matching data
void list_delete(list_t *list, ll_data_t data) {
  if (!list || list_is_empty(list)) {
    return;
  }
  for (list_node_t *cur = list->first; cur; cur = cur->next) {
    if (cur->data != data) {
      continue;
    }

    list->count--;
    list_node_t **prev = cur->prev ? &cur->prev->next : &list->first;
    list_node_t **next = cur->next ? &cur->next->prev : &list->last;
    *prev = cur->next;
    *next = cur->prev;
    node_destroy(cur);
    return;
  }
}

// constructs a new (empty) list
list_t *list_create() {
  list_t *list = calloc(1, sizeof(list_t));
  return list;
}

// destroys an entire list
// list will be a dangling pointer after calling this method on it
void list_destroy(list_t *list) {
  if (!list) {
    return;
  }

  for (list_node_t *cur = list->first; cur; cur = list->first) {
    list->first = cur->next;
    node_destroy(cur);
  }
  free(list);
}

// Check if the list is empty.
bool list_is_empty(const list_t *list) { return !list->first; }

// Checks if the list has only one element
bool list_is_single(const list_t *list) {
  return list->first == list->last && list->first;
}

// Create a new node that stores the passed value and has the given neighbors
list_node_t *node_create(ll_data_t data, list_node_t *prev, list_node_t *next) {
  list_node_t *node = malloc(sizeof(list_node_t));

  node->data = data;
  node->prev = prev;
  node->next = next;

  return node;
}

// Destroy the past without destroying your neighbors
void node_destroy(list_node_t *node) {
  if (node) {
    free(node);
  }
}

File: go/linked-list/linked_list.go
// Package linkedlist provides an implementation of a doubly linked list.
// The list can store elements of any type using the empty interface.
package linkedlist

import "errors"

// List represents a doubly linked list.
type List struct {
	first *Node
	last  *Node
}

// NewList creates and returns a new List initialized with the provided elements.
// The elements are added to the end of the list in the order they are given.
func NewList(elements ...interface{}) *List {
	list := List{}

	for _, element := range elements {
		list.Push(element)
	}

	return &list
}

// First returns a pointer to the first Node of the list.
// It returns nil if the list is empty.
func (l *List) First() *Node {
	return l.first
}

// Last returns a pointer to the last Node of the list.
// It returns nil if the list is empty.
func (l *List) Last() *Node {
	return l.last
}

// Unshift adds a new element to the beginning of the list.
func (l *List) Unshift(v interface{}) {
	if l.isEmpty() {
		l.insertInEmpty(v)
		return
	}
	l.first.prev = newNode(v, nil, l.first)
	l.first = l.first.prev
}

// Push adds a new element to the end of the list.
func (l *List) Push(v interface{}) {
	if l.isEmpty() {
		l.insertInEmpty(v)
		return
	}
	l.last.next = newNode(v, l.last, nil)
	l.last = l.last.next
}

// Shift removes and returns the first element of the list.
// It returns an error if the list is empty.
func (l *List) Shift() (interface{}, error) {
	if l.isEmpty() {
		return nil, errors.New("can't shift from a empty list")
	}
	if l.isSingle() {
		return l.removeFromSingle(), nil
	}
	popped := l.first.Value

	l.first = l.first.next
	l.first.prev = nil

	return popped, nil
}

// Pop removes and returns the last element of the list.
// It returns an error if the list is empty.
func (l *List) Pop() (interface{}, error) {
	if l.isEmpty() {
		return nil, errors.New("can't pop from a empty list")
	}
	if l.isSingle() {
		return l.removeFromSingle(), nil
	}
	popped := l.last.Value

	l.last = l.last.prev
	l.last.next = nil

	return popped, nil
}

// Reverse reverses the order of the elements in the list in place.
func (l *List) Reverse() {
	var next *Node = nil

	for current := l.first; current != nil; current = next {
		next = current.next
		current.prev, current.next = current.next, current.prev
	}

	l.first, l.last = l.last, l.first
}

// isEmpty checks if the list is empty.
func (l *List) isEmpty() bool {
	return l.first == nil
}

// isSingle checks if the list contains only one node.
func (l *List) isSingle() bool {
	return !l.isEmpty() && l.first == l.last
}

// insertInEmpty is a helper method for adding the first node to an empty list.
func (l *List) insertInEmpty(value interface{}) {
	l.first = newNode(value, nil, nil)
	l.last = l.first
}

// removeFromSingle is a helper method for removing the only node from a list.
func (l *List) removeFromSingle() interface{} {
	value := l.last.Value
	l.first = nil
	l.last = nil
	return value
}

// Node represents a node in the doubly linked list.
type Node struct {
	Value interface{}
	prev  *Node
	next  *Node
}

// Next returns the next Node in the list.
func (n *Node) Next() *Node {
	return n.next
}

// Prev returns the previous Node in the list.
func (n *Node) Prev() *Node {
	return n.prev
}

// newNode is a helper function to create a new Node with the given value and links.
func newNode(value interface{}, prev, next *Node) *Node {
	return &Node{Value: value, prev: prev, next: next}
}
